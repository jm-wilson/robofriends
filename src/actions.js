import {
  CHANGE_SEARCH_FIELD,
  REQUEST_ROBOTS_PENDING,
  REQUEST_ROBOTS_SUCCESS,
  REQUEST_ROBOTS_FAILED,
  FILTER_ROBOTS,
  CHANGE_CLICKME_COUNTER,
} from './constants.js';

export const setSearchField = (text) => ({
  type: CHANGE_SEARCH_FIELD,
  payload: text,
});

export const requestRobots = (dispatch) => {
  dispatch({ type: REQUEST_ROBOTS_PENDING });
  fetch('https://jsonplaceholder.typicode.com/users')
    .then((response) => response.json())
    .then((data) => {
      dispatch({ type: REQUEST_ROBOTS_SUCCESS, payload: data });
      dispatch({ type: FILTER_ROBOTS, payload: data });
    })
    .catch((error) => dispatch({ type: REQUEST_ROBOTS_FAILED, payload: error }));
};

export const filterRobots = (text, robots) => {
  const payload = robots.filter((robot) => {
    return robot.name.toLowerCase().includes(text.toLowerCase());
  });

  return {
    type: FILTER_ROBOTS,
    payload: payload,
  };
};

export const incrementClickme = (counter) => ({
  type: CHANGE_CLICKME_COUNTER,
  payload: counter + 1,
});
