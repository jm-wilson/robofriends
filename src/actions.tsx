import {
  CHANGE_SEARCH_FIELD,
  REQUEST_ROBOTS_PENDING,
  REQUEST_ROBOTS_SUCCESS,
  REQUEST_ROBOTS_FAILED,
  FILTER_ROBOTS,
  CHANGE_CLICKME_COUNTER,
} from './constants';
import { IRobot } from './containers/App';
import { Dispatch } from 'redux';

export interface ActionWithStringPayload {
  type?: string;
  payload?: string;
}
export interface ActionWithObjectPayload {
  type?: string;
  payload?: object;
}
export interface ActionWithNumberPayload {
  type?: string;
  payload?: number;
}

export const setSearchField = (text: string): ActionWithStringPayload => ({
  type: CHANGE_SEARCH_FIELD,
  payload: text,
});

export const requestRobots =
  (apiLink: string = 'https://jsonplaceholder.typicode.com/users') =>
  async (dispatch: Dispatch) => {
    try {
      dispatch({ type: REQUEST_ROBOTS_PENDING });
      const data = await fetch(apiLink).then((response) => response.json());
      dispatch({ type: REQUEST_ROBOTS_SUCCESS, payload: data });
      dispatch({ type: FILTER_ROBOTS, payload: data });
    } catch (error) {
      dispatch({ type: REQUEST_ROBOTS_FAILED, payload: error });
    }
  };

export const filterRobots = (text: string, robots: IRobot[]): ActionWithObjectPayload => {
  const payload = robots.filter((robot) => {
    return robot.name.toLowerCase().includes(text.toLowerCase());
  });

  return {
    type: FILTER_ROBOTS,
    payload: payload,
  };
};

export const incrementClickme = (counter: number): ActionWithNumberPayload => ({
  type: CHANGE_CLICKME_COUNTER,
  payload: counter + 1,
});
