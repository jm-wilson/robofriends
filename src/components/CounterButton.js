import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { incrementClickme } from '../actions';

const CounterButton = () => {
  const dispatch = useDispatch();
  const counter = useSelector((state) => state.clickMe.counter);

  const onClickMe = () => {
    dispatch(incrementClickme(counter));
  };

  return (
    <button id='btn-counter' onClick={onClickMe}>
      Click me ({counter})
    </button>
  );
};

export default CounterButton;
