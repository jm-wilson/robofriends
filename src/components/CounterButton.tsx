import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { incrementClickme } from '../actions';
import { IRootState } from '../containers/App';

const CounterButton: React.FC = (): JSX.Element => {
  const dispatch = useDispatch();
  const counter: number = useSelector((state: IRootState) => state.clickMe.counter);

  const onClickMe = () => {
    dispatch(incrementClickme(counter));
  };

  return (
    <button id='btn-counter' onClick={onClickMe}>
      Click me ({counter})
    </button>
  );
};

export default CounterButton;
