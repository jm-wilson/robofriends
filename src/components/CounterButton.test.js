import React from 'react';
import { shallow } from 'enzyme';
import CounterButton from './CounterButton';
import { useSelector, useDispatch } from 'react-redux';
import { incrementClickme } from '../actions';

const mockDispatch = jest.fn();

test('should say "Click me (0)" when counter is 0', () => {
  useSelector.mockImplementation(() => 0);

  expect(shallow(<CounterButton />)).toMatchSnapshot();
});

test('should say "Click me (5)" when counter is 5', () => {
  useSelector.mockImplementation(() => 5);

  expect(shallow(<CounterButton />)).toMatchSnapshot();
});

test('should run useDispatch on click', () => {
  useDispatch.mockImplementation(() => mockDispatch);

  const wrapper = shallow(<CounterButton />);

  wrapper.find('[id="btn-counter"]').simulate('click');
  expect(mockDispatch).toBeCalledTimes(1);
  expect(mockDispatch).toBeCalledWith(incrementClickme());
});
