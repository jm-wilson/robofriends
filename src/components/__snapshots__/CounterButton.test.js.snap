// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should say "Click me (0)" when counter is 0 1`] = `
<button
  id="btn-counter"
  onClick={[Function]}
>
  Click me (
  0
  )
</button>
`;

exports[`should say "Click me (5)" when counter is 5 1`] = `
<button
  id="btn-counter"
  onClick={[Function]}
>
  Click me (
  5
  )
</button>
`;
